proxywasm/proxytest/wasmwrapper.go

```diff
@@ -0,0 +1,423 @@
+	handleErr(err)
+	return res[0] == 1
+}
+
+func (p *pluginContext) OnQueueReady(queueID uint32) {
+	_, err := p.abi.proxyOnQueueReady.Call(p.ctx, p.id, uint64(queueID))
+	handleErr(err)
+}
+
+func (p *pluginContext) OnTick() {
+	_, err := p.abi.proxyOnTick.Call(p.ctx, p.id)
+	handleErr(err)
+}
+
+func (p *pluginContext) NewTcpContext(uint32) types.TcpContext {
+	return nil
+}
+
+func (p *pluginContext) NewHttpContext(contextID uint32) types.HttpContext {
+	_, err := p.abi.proxyOnContextCreate.Call(p.ctx, uint64(contextID), p.id)
+	handleErr(err)
+	return &httpContext{
+		id:  uint64(contextID),
+		abi: p.abi,
+		ctx: p.ctx,
+	}
+}
+
+type httpContext struct {
+	id  uint64
+	abi guestABI
+	ctx context.Context
+}
+
+func (h *httpContext) OnHttpRequestHeaders(numHeaders int, endOfStream bool) types.Action {
+	res, err := h.abi.proxyOnRequestHeaders.Call(h.ctx, h.id, uint64(numHeaders), wasmBool(endOfStream))
+	handleErr(err)
+	return types.Action(res[0])
+}
+
+func (h *httpContext) OnHttpRequestBody(bodySize int, endOfStream bool) types.Action {
+	res, err := h.abi.proxyOnRequestBody.Call(h.ctx, h.id, uint64(bodySize), wasmBool(endOfStream))
+	handleErr(err)
+	return types.Action(res[0])
+}
+
+func (h *httpContext) OnHttpRequestTrailers(numTrailers int) types.Action {
+	res, err := h.abi.proxyOnRequestTrailers.Call(h.ctx, h.id, uint64(numTrailers))
+	handleErr(err)
+	return types.Action(res[0])
+}
+
+func (h *httpContext) OnHttpResponseHeaders(numHeaders int, endOfStream bool) types.Action {
+	res, err := h.abi.proxyOnResponseHeaders.Call(h.ctx, h.id, uint64(numHeaders), wasmBool(endOfStream))
+	handleErr(err)
+	return types.Action(res[0])
+}
+
+func (h *httpContext) OnHttpResponseBody(bodySize int, endOfStream bool) types.Action {
+	res, err := h.abi.proxyOnResponseBody.Call(h.ctx, h.id, uint64(bodySize), wasmBool(endOfStream))
+	handleErr(err)
+	return types.Action(res[0])
+}
+
+func (h *httpContext) OnHttpResponseTrailers(numTrailers int) types.Action {
+	res, err := h.abi.proxyOnResponseTrailers.Call(h.ctx, h.id, uint64(numTrailers))
+	handleErr(err)
+	return types.Action(res[0])
+}
+
+func (h *httpContext) OnHttpStreamDone() {
+	_, err := h.abi.proxyOnLog.Call(h.ctx, h.id)
+	handleErr(err)
+}
+
+func handleErr(err error) {
+	if err != nil {
+		panic(err)
+	}
+}
+
+func handleMemoryStatus(ok bool) {
+	if !ok {
+		panic("could not access memory")
+	}
+}
+
+func wasmBytePtr(ctx context.Context, mod api.Module, off uint32, size uint32) *byte {
+	if size == 0 {
+		return nil
+
+	}
+	buf, ok := mod.Memory().Read(ctx, off, size)
+	handleMemoryStatus(ok)
+	return &buf[0]
+}
+
+func copyBytesToWasm(ctx context.Context, mod api.Module, hostPtr *byte, size int, wasmPtrPtr uint32, wasmSizePtr uint32) {
+	var hostSlice []byte
+	hdr := (*reflect.SliceHeader)(unsafe.Pointer(&hostSlice))
+	hdr.Data = uintptr(unsafe.Pointer(hostPtr))
+	hdr.Cap = size
+	hdr.Len = size
+
+	// proxy-wasm only defines alloc without free so this leaks memory within the wasm module.
```

Comment:
Couldn't come up with a good solution for this

Reply:
strictly speaking it's not leaking as GC tries to deallocate if it runs of the memory space

Reply:
TinyGo reads the memory here

https://github.com/tetratelabs/proxy-wasm-go-sdk/blob/main/proxywasm/internal/hostcall_utils_tinygo.go#L37

AFAIK, because of the use of slice literal construction `reflect.SliceHeader{}`, rather than casting a variable of slice type, this variable is not tracked by GC (unless it behaves differently in TinyGo vs Go). Do you mean it would still free allocations regardless of reference tracking?

Reply:
> this variable is not tracked by GC

I would be surprised if it's not -- TinyGo's GC tried to mark all the possible objects in use, and this case the byte slice returned by `RawBytePtrToByteSlice` should be marked as in-use eventually and tracked

Reply:
But I think we should do some experiments on this -- and if this is leaking, I would add global map to track allocations in memory

Reply:
Ah ok I think that's right. Then there shouldn't be any leak here, cool